ARCs/arc-0020.md

```diff
@@ -0,0 +1,870 @@
+---
+arc: 0020
+title: Smart ASA
+description: An ARC for an ASA controlled by an Algorand Smart Contract
+author:
+  - Cosimo Bassi (@cusma)
+  - Adriano Di Luzio (@aldur)
+  - John Jannotti (@jannotti)
+discussions-to: <URL>
+status: Draft
+type: Standards Track
+category (*only required for Standards Track): Interface
+created: 2022-04-27
+requires: ARC-4
+---
+
+## Abstract
+
+A "Smart ASA" is an Algorand Standard Asset (ASA) controlled by a Smart Contract
+that exposes methods to create, configure, transfer, freeze, and destroy the
+asset.
+
+This ARC defines the ABI interface of such Smart Contract, the required
+metadata, and suggests a reference implementation.
+
+## Motivation
+
+The Algorand Standard Asset (ASA) is an excellent building block for on-chain
+applications. It is battle-tested and widely supported by SDKs, wallets, and
+dApps.
+
+However, the ASA lacks in flexibility and configurability. For instance, once
+issued it can't be re-configured (its unit name, decimals, maximum supply).
+Also, it is freely transferable (unless frozen). This prevents developers from
+specifying additional business logic to be checked while transferring it (think
+of royalties or [vesting](https://en.wikipedia.org/wiki/Vesting)).
+
+Enforcing transfer conditions require freezing the asset and transferring it
+through a clawback operation -- which results in a process that is opaque to
+users and wallets and a bad experience for the users.
+
+The Smart ASA defined by this ARC extends the ASA to increase its expressiveness
```

Comment:
Minor suggestion (speaks to the purpose of it being a standard and the flow on effect of having confidence supporting it across the ecosystem - people could do this today, but nothing supports it).

From:

> Developers can now introduce additional business logic around its operations. Wallets, dApps, and SDKs can recognize Smart ASAs and adjust their flows and user experiences accordingly.

To:

> By introducing this as a standard, marketplaces, wallets, dApps, and SDKs can confidently and consistently recognize Smart ASAs and adjust their flows and user experiences accordingly.

Reply:
To:

> By introducing this as a standard both developers, users (marketplaces, wallets, dApps, etc.) and SDKs can confidently and consistently recognize Smart ASAs and adjust their flows and user experiences accordingly.